### ISSUE TITLE
Clarification on client capabilities

### ISSUE THREAD

**Issue by ljw1004** [Microsoft/language-server-protocol#193]
The spec says this about servers receiving a ClientCapabilities objet literal: *"A missing property should be interpreted as an absence of the capability."*

In several places it's not clear precisely which capability is in question. I'd love clarification, please.

```
WorkspaceClientCapabilites
  applyEdit?: boolean; // the client supports applying batch edits
```
* Does "applying batch edits" capability refer specifically to whether the client supports having the server send it a `workspace/applyEdits` request?
* Or does it more generally say whether the client can handle a batch of edits -- i.e. every `TextDocumentEdit` it receives must contain no more than one edit, and likewise every `textDocument/willSaveUntilTextDocument`, and likewise `compltionItem.additionalTextEdits` must be empty, and likewise the response to `textDocument/formatting` must be a singleton edit?

```
TextDocumentClientCapabilities
    synchronization?: {
        dynamicRegistration?: boolean;
        willSave?: boolean;
        willSaveWaitUntil?: boolean;
        didSave?: boolean; }
```
* If the entire `synchronization` field is absent, does that mean the client declines to do any kind of synchronization at all (i.e. won't send the didChange notification) and so expects the server to tell it to use TextDocumentSyncKind.None?
* Or does a missing `synchronization` field merely mean we should assume the four inner capabilities are false?

```
WorkspaceClientCapabilities {
  workspaceEdit?: {
    documentChanges?: boolean; // client supports versioned doc changes
```
* If the entire workspaceEdit object is absent, does this mean the client doesn't support anything that involves workspaceEdit records, i.e. it doesn't support `textDocument/rename` and `workspace/applyEdit`?
* Or does it merely mean that `documentChanges = false`?

```
WorkspaceClientCapabilities {
  symbol?: {
    dynamicRegistration?: boolean;
```
* If the entire `symbol` field is absent, it might mean that the client will decline to ever send the `workspace/Symbol` method.
* Or it might simple mean to assume `symbol.dynamicRegistration = false`. I guess it's the latter (since there's no value in having a client tell the server which messages it will or won't send).

In exactly this same bucket are a load of other capabilities:
```
didChangeConfiguration?: {...}
didChangeWatchedFiles?: {...}
executeCommand?: {...}
completion?: {...}
hover?: {...}
signatureHelp?: {...}
references?: {...}
documentHighlight?: {...}
documentSymbol?: {...}
formatting?: {...}
rangeFormatting?: {...}
onTypeFormatting?: {...}
definition?: {...}
codeAction?: {...}
codeLens?: {...}
documentLink?: {...}
rename?: {...}
```

**Comment by ljw1004** [comment]
Another one:

```
TextDocumentSyncOptions
  save?: {
    includeText?: boolean // The client is supposed to include the content on save
```
* If the entire "save" object is absent, should the client omit the optional `text` field when it sends `textDocument/didSave` notifications?
* Or should it refrain from sending `didSave` notifications at all?

**Comment by dbaeumer** [comment]
I added a clarification what should happen in the case of sub properties.

I also add clarification how client capabilities should be interpreted in general. The statement is:

Client capabilities got introduced with the version 3.0 of the protocol. They therefore only describe capabilities that got introduced in 3.x or later. Capabilities that existed in the 2.x version of the protocol are still mandatory for clients. Clients can't not opt out of providing them. So even if a client omits the `ClientCapabilities.textDocument.synchronization` it is still required that the client provides text document synchronization (e.g. open, changed and close notifications).

Regarding applyEdit. This refers to the client request workspace/applyEdit. Added clarification as well.
